src/com/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -928,119 +757,269 @@
   "Given a certname, querys the DB for existing facts for that
    certname and will update, delete or insert the facts as necessary
-   to match the facts argument."
-  [{:keys [name values environment timestamp producer-timestamp] :as fact-data} :- facts-schema]
-  (let [factset (certname-to-factset-id name)
-        old-facts (current-fact-value-ids factset)
-        new-facts (new-fact-value-ids values)]
-    (sql/update-values :factsets ["id=?" factset]
-                       {:timestamp (to-timestamp timestamp)
-                        :environment_id (ensure-environment environment)
-                        :producer_timestamp (to-timestamp producer-timestamp)})
-    (utils/diff-fn (zipmap new-facts (repeat nil))
-                   (zipmap old-facts (repeat nil))
-                   #(insert-facts! factset %)
-                   #(delete-facts! factset %)
-                   identity)))
+   to match the facts argument. (cf. add-facts!)"
+  [{:keys [name values environment timestamp producer-timestamp] :as fact-data}]
+
+  ;; Might be possible to make some of this faster via sorted parallel traversal.
+  (sql/transaction
+   (let [factset-id (certname-to-factset-id name)
+         initial-factset-paths-vhashes
+         (sql/with-query-results result-set
+           ["SELECT fp.path, fv.value_hash FROM facts f
+               INNER JOIN fact_paths fp ON f.fact_path_id = fp.id
+               INNER JOIN fact_values fv ON f.fact_value_id = fv.id
+               WHERE factset_id = ?"
+            factset-id]
+           (doall result-set))
+         ;; Ensure that all the required paths and values exist.
+         paths-and-valuemaps (facts/facts->paths-and-valuemaps values)
+         pathstrs (map (comp facts/factpath-to-string first) paths-and-valuemaps)
+         path-ids (map (realize-paths! pathstrs) pathstrs)
+         valuemaps (map second paths-and-valuemaps)
+         vhashes (map :value_hash valuemaps)
+         vhash-ids (map :id (map (realize-values! valuemaps) vhashes))]
+
+     ;; Add new facts and remove obsolete facts.
+     (let [replacement-pv-pairs (set (map vector path-ids vhash-ids))
+           current-pairs (set (select-pid-vid-pairs-for-factset factset-id))
+           new-pairs (set/difference replacement-pv-pairs current-pairs)
+           rm-pairs (set/difference current-pairs replacement-pv-pairs)]
+
+       (when-let [rm-pairs (seq rm-pairs)]
+         (sql/do-prepared
+          (format "DELETE FROM facts
+                   WHERE factset_id = ? AND (fact_path_id, fact_value_id) %s"
+                  (jdbc/in-clause-multi rm-pairs 2))
+          ;; CAUTION: seq required - currently: (flatten #{[1 2] [3 4]}) => ()
+          (flatten [factset-id rm-pairs])))
+
+       (insert-facts-pv-pairs! factset-id new-pairs)
+
+       (when-let [rm-pairs (seq rm-pairs)]
+         (delete-pending-path-id-orphans! factset-id
+                                          (set/difference (set rm-pairs)
```

Comment:
@senior I already changed this during the rework to take advantage of factset-id/path uniqueness.